Generation fails for expect declared in intermediate source set

Create a KMP project with jvm + linuxX64 + macosX64. And such code in source sets:
```Kotlin
// nativeMain
expect fun shared()

// linuxX64Main, macosX64Main
actual fun shared(){}
```


```
Caused by: java.lang.IllegalArgumentException: Collection has more than one element.
	at kotlin.collections.CollectionsKt___CollectionsKt.single(_Collections.kt:605)
	at org.jetbrains.dokka.transformers.documentation.DefaultDocumentableMerger.mergeExpectActual$analyzeExpectActual(DefaultDocumentableMerger.kt:119)
	at org.jetbrains.dokka.transformers.documentation.DefaultDocumentableMerger.mergeExpectActual(DefaultDocumentableMerger.kt:136)
	at org.jetbrains.dokka.transformers.documentation.DefaultDocumentableMerger.mergeWith(DefaultDocumentableMerger.kt:141)
	at org.jetbrains.dokka.transformers.documentation.DefaultDocumentableMerger$invoke$1$3.invoke(DefaultDocumentableMerger.kt:35)
	at org.jetbrains.dokka.transformers.documentation.DefaultDocumentableMerger$invoke$1$3.invoke(DefaultDocumentableMerger.kt:33)
	at org.jetbrains.dokka.transformers.documentation.DefaultDocumentableMerger.merge(DefaultDocumentableMerger.kt:55)
	at org.jetbrains.dokka.transformers.documentation.DefaultDocumentableMerger.invoke(DefaultDocumentableMerger.kt:33)
	at org.jetbrains.dokka.base.generation.SingleModuleGeneration.mergeDocumentationModels(SingleModuleGeneration.kt:82)
	at org.jetbrains.dokka.base.generation.SingleModuleGeneration.generate(SingleModuleGeneration.kt:38)
	at org.jetbrains.dokka.DokkaGenerator$generate$1.invoke(DokkaGenerator.kt:34)
	at org.jetbrains.dokka.DokkaGenerator$generate$1.invoke(DokkaGenerator.kt:27)
	at org.jetbrains.dokka.DokkaGeneratorKt.timed(DokkaGenerator.kt:82)
	at org.jetbrains.dokka.DokkaGeneratorKt.access$timed(DokkaGenerator.kt:1)
	at org.jetbrains.dokka.DokkaGenerator.generate(DokkaGenerator.kt:27)
	at org.jetbrains.dokka.DokkaBootstrapImpl.generate(DokkaBootstrapImpl.kt:67)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.jetbrains.dokka.gradle.DelegatedInvocationHandler.invoke(automagicTypedProxy.kt:44)
	at jdk.proxy4/jdk.proxy4.$Proxy127.generate(Unknown Source)
	at org.jetbrains.dokka.gradle.AbstractDokkaTask$generateDocumentation$1$1.run(AbstractDokkaTask.kt:232)
```

**Installation**
- Dokka version: 2.0.20-dev-357
- DGPv1 / K1 or K2

[dokka-repros.zip](https://github.com/user-attachments/files/16966581/dokka-repros.zip)

--- a/dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/transformers/documentation/DefaultDocumentableMerger.kt
+++ b/dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/transformers/documentation/DefaultDocumentableMerger.kt
@@ -116,6 +116,7 @@ public class DefaultDocumentableMerger(context: DokkaContext) : DocumentableMerg
                 listOf(actuals to actuals.flatMap { it.sourceSets }.toSet())
             } else expects.map { expect ->
                 val actualsForGivenExpect = actuals.filter { actual ->
+                    // [actual.sourceSets] can already be partially merged and contain more than one source set
                     actual.sourceSets.all {
                         dependencyInfo[it]
                         ?.contains(expect.expectPresentInSet!!)

--- a/dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/transformers/documentation/DefaultDocumentableMerger.kt
+++ b/dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/transformers/documentation/DefaultDocumentableMerger.kt
@@ -116,9 +116,11 @@ public class DefaultDocumentableMerger(context: DokkaContext) : DocumentableMerg
                 listOf(actuals to actuals.flatMap { it.sourceSets }.toSet())
             } else expects.map { expect ->
                 val actualsForGivenExpect = actuals.filter { actual ->
-                    dependencyInfo[actual.sourceSets.single()]
+                    actual.sourceSets.all {
+                        dependencyInfo[it]
                         ?.contains(expect.expectPresentInSet!!)
                         ?: throw IllegalStateException("Cannot resolve expect/actual relation for ${actual.name}")
+                    }
                 }
                 (listOf(expect) + actualsForGivenExpect) to actualsForGivenExpect.flatMap { it.sourceSets }.toSet()
             }

--- a/dokka-subprojects/plugin-base/src/test/kotlin/expectActuals/ExpectActualsTest.kt
+++ b/dokka-subprojects/plugin-base/src/test/kotlin/expectActuals/ExpectActualsTest.kt
@@ -4,6 +4,7 @@
 
 package expectActuals
 
+import org.jetbrains.dokka.DokkaSourceSetID
 import org.jetbrains.dokka.base.testApi.testRunner.BaseAbstractTest
 import org.jetbrains.dokka.model.withDescendants
 import org.jetbrains.dokka.pages.ClasslikePageNode
@@ -378,6 +379,60 @@ class ExpectActualsTest : BaseAbstractTest() {
                 setOf("jvm", "native"),
                 classlike.sourceSets.map { it.sourceSetID.sourceSetName }.toSet()
             )
+        }
+    }
+
+
+    @Test
+    fun `should work with the reversed order of source sets #3798`() {
+        val configuration = dokkaConfiguration {
+            sourceSets {
+                val commonId = DokkaSourceSetID("root", "common")
+                sourceSet {
+                    sourceRoots = listOf("src/jvm/")
+                    analysisPlatform = "jvm"
+                    name = "jvm"
+                    displayName = "jvm"
+                    dependentSourceSets = setOf(commonId)
+                }
+                sourceSet {
+                    sourceRoots = listOf("src/native/")
+                    analysisPlatform = "native"
+                    name = "native"
+                    displayName = "native"
+                    dependentSourceSets = setOf(commonId)
+                }
+                sourceSet {
+                    sourceRoots = listOf("src/common/")
+                    analysisPlatform = "common"
+                    name = "common"
+                    displayName = "common"
+                }
+            }
+        }
+
+        testInline(
+            """
+        /src/common/test.kt
+        expect fun shared()
+        
+        /src/native/test.kt
+        actual fun shared(){}
+        
+        /src/jvm/test.kt
+        actual fun shared(){}
+        """.trimMargin(), configuration
+        ) {
+            documentablesTransformationStage = { module ->
+                val function = module.packages.single().functions.single { it.name == "shared" }
+                assertTrue(function.isExpectActual)
+                assertEquals(
+                    "common", function.expectPresentInSet?.sourceSetID?.sourceSetName
+                )
+                assertEquals(
+                    setOf("common", "jvm", "native"), function.sourceSets.map { it.sourceSetID.sourceSetName }.toSet()
+                )
+            }
         }
     }
 }

